Update debian java dependency

spacecowboy opened the pull request:
This is a backport and update of the fix for #7188

It does not use the `Conflicts` line we previously added to prevent the
installation of openjdk-9 on Ubuntu 16.04. (when forward merging should
take care to remove that line from the 3.x branches, and update version
to java-8 as well).

The relevant part of `Depends:` is:

```
openjdk-N-jre | javaN-runtime
```

(where N will be 7 for 2.x and 8 for 3.x)

`javaN-runtime` is a virtual package which for example `openjdk-N-jre`
provides. Installing for example an Oracle JRE from webupd8 will also
provide it.

The line has the following effect:
- If neither package is found/provided on the system, it will install
  `openjdk-N-jre`
- If either of the packages is found/provided on the system, no
  additional JRE will be installed.
## Tests to be verified:
### Neo4j 2.x
- [x] debian 7/8
- [x] ubuntu 14.04
- [x] ubuntu 16.04
- [x] ubuntu 14.04 (with pre installed jre)
- [x] ubuntu 16.04 (with pre installed jre)
### Upgrade Neo4j from 2.x to 3.0
- [x] debian 8
- [x] ubuntu 14.04
- [x] ubuntu 16.04
## Some simple testing of this behavior on a fresh Ubuntu 16.04 docker instance follows.

First setup the webupd8 ppa to be able to test with another JRE:

```
apt-get update && \
  apt-get install -y software-properties-common && \
  add-apt-repository ppa:webupd8team/java && \
  apt-get update
```

Then, install neo4j (example uses this change but built for 3.1)

```
dpkg -i neo4j_3.1.0.SNAPSHOT_all.deb
```

dpkg will complain about missing packages

```
Selecting previously unselected package neo4j.
(Reading database ... 10127 files and directories currently installed.)
Preparing to unpack neo4j_3.1.0.SNAPSHOT_all.deb ...
Unpacking neo4j (3.1.0.SNAPSHOT) ...
dpkg: dependency problems prevent configuration of neo4j:
 neo4j depends on daemon; however:
  Package daemon is not installed.
 neo4j depends on psmisc; however:
  Package psmisc is not installed.
 neo4j depends on openjdk-8-jre-headless | java8-runtime; however:
  Package openjdk-8-jre-headless is not installed.
  Package java8-runtime is not installed.

dpkg: error processing package neo4j (--install):
 dependency problems - leaving unconfigured
Processing triggers for systemd (229-4ubuntu4) ...
Errors were encountered while processing:
 neo4j
```

Fix them and see that
openjdk-9 is NOT installed

```
apt-get install -f
```

```
The following NEW packages will be installed:
  ca-certificates-java daemon fontconfig-config fonts-dejavu-core java-common libavahi-client3 libavahi-common-data
  libavahi-common3 libcups2 libfontconfig1 libfreetype6 libjpeg-turbo8 libjpeg8 liblcms2-2 libnspr4 libnss3 libnss3-nssdb
  libpcsclite1 libpng12-0 libx11-6 libx11-data libxau6 libxcb1 libxdmcp6 libxext6 libxi6 libxrender1 libxtst6
  openjdk-8-jre-headless psmisc x11-common
```

Now, we try with Oracle JRE instead

Uninstall neo4j and java again

```
apt-get purge neo4j openjdk-8-jre-headless
```

Install Oracle JRE

```
apt-get install oracle-java8-installer
```

and install neo4j and notice that nothing is missing and nothing more is
installed

```
dpkg -i neo4j_3.1.0.SNAPSHOT_all.deb
```

```
Selecting previously unselected package neo4j.
(Reading database ... 11232 files and directories currently installed.)
Preparing to unpack neo4j_3.1.0.SNAPSHOT_all.deb ...
Unpacking neo4j (3.1.0.SNAPSHOT) ...
Setting up neo4j (3.1.0.SNAPSHOT) ...
invoke-rc.d: policy-rc.d denied execution of start.
Processing triggers for systemd (229-4ubuntu4) ...
```

spacecowboy committed:
Update debian java dependency

This is a backport and update of the fix for #7188

It does not use the `Conflicts` line we previously added to prevent the
installation of openjdk-9 on Ubuntu 16.04. (when forward merging should
take care to remove that line from the 3.x branches, and update version
to java-8 as well).

The relevant part of `Depends:` is:

```
openjdk-N-jre-headless | javaN-runtime-headless
```

(where N will be 7 for 2.x and 8 for 3.x)

`javaN-runtime` is a virtual package which for example `openjdk-N-jre`
provides. Installing for example an Oracle JRE from webupd8 will also
provide it.

The line has the following effect:

- If neither package is found/provided on the system, it will install
  `openjdk-N-jre-headless`
- If either of the packages is found/provided on the system, no
  additional JRE will be installed.

Some simple testing of this behavior on a fresh Ubuntu 16.04 docker
instance follows.

First setup the webupd8 ppa to be able to test with another JRE:

```
apt-get update && \
  apt-get install -y software-properties-common && \
  add-apt-repository ppa:webupd8team/java && \
  apt-get update
```

Then, install neo4j (example uses this change but built for 3.1)

```
dpkg -i neo4j_3.1.0.SNAPSHOT_all.deb
```

dpkg will complain about missing packages

```
Selecting previously unselected package neo4j.
(Reading database ... 10127 files and directories currently installed.)
Preparing to unpack neo4j_3.1.0.SNAPSHOT_all.deb ...
Unpacking neo4j (3.1.0.SNAPSHOT) ...
dpkg: dependency problems prevent configuration of neo4j:
 neo4j depends on daemon; however:
  Package daemon is not installed.
 neo4j depends on psmisc; however:
  Package psmisc is not installed.
 neo4j depends on openjdk-8-jre-headless | java8-runtime; however:
  Package openjdk-8-jre-headless is not installed.
  Package java8-runtime is not installed.

dpkg: error processing package neo4j (--install):
 dependency problems - leaving unconfigured
Processing triggers for systemd (229-4ubuntu4) ...
Errors were encountered while processing:
 neo4j
```

Fix them and see that
openjdk-9 is NOT installed

```
apt-get install -f
```

```
The following NEW packages will be installed:
  ca-certificates-java daemon fontconfig-config fonts-dejavu-core java-common libavahi-client3 libavahi-common-data
  libavahi-common3 libcups2 libfontconfig1 libfreetype6 libjpeg-turbo8 libjpeg8 liblcms2-2 libnspr4 libnss3 libnss3-nssdb
  libpcsclite1 libpng12-0 libx11-6 libx11-data libxau6 libxcb1 libxdmcp6 libxext6 libxi6 libxrender1 libxtst6
  openjdk-8-jre-headless psmisc x11-common
```

Now, we try with Oracle JRE instead

Uninstall neo4j and java again

```
apt-get purge neo4j openjdk-8-jre-headless
```

Install Oracle JRE

```
apt-get install oracle-java8-installer
```

and install neo4j and notice that nothing is missing and nothing more is
installed

```
dpkg -i neo4j_3.1.0.SNAPSHOT_all.deb
```

```
Selecting previously unselected package neo4j.
(Reading database ... 11232 files and directories currently installed.)
Preparing to unpack neo4j_3.1.0.SNAPSHOT_all.deb ...
Unpacking neo4j (3.1.0.SNAPSHOT) ...
Setting up neo4j (3.1.0.SNAPSHOT) ...
invoke-rc.d: policy-rc.d denied execution of start.
Processing triggers for systemd (229-4ubuntu4) ...
```

benbc commented:
@spacecowboy This looks like a great improvement -- thanks. 

Please can you consider how this behaves for 3.x on systems that don't provide Java 8 in the repos (e.g. Ubuntu 14.04). I think it will be fine since it's analogous to the Oracle JRE example you've tested, but it's worth checking.

Please also test upgrades.

spacecowboy committed:
Add alternate dependency openjdk-8

Because Ubuntu 16.04 no longer has openjdk-7, so it will fallback to
install openjdk-8 now (previously, it would pick openjdk-9 which is all
kinds of wrong).

and keep depending on the non-headless variant so we don't change too
much in a patch release.

spacecowboy commented:
- Debian 7/8

```
The following extra packages will be installed:
  ....
  openjdk-7-jre openjdk-7-jre-headless
  ....
```
- Ubuntu 14.04

```
The following extra packages will be installed:
  ....
  openjdk-7-jre openjdk-7-jre-headless
  ....
```
- Ubuntu 16.04

```
The following extra packages will be installed:
  ....
  openjdk-8-jre openjdk-8-jre-headless
  ....
```
- Ubuntu 14.04 with Oracle Java7 already installed

```
The following NEW packages will be installed:
  daemon lsof psmisc
```
- Ubuntu 16.04 with Oracle Java7 already installed

```
The following NEW packages will be installed:
  daemon lsof psmisc
```

spacecowboy commented:
Note that in 3.x, the line(s) should read:

```
Depends: ${misc:Depends}, bash, daemon, adduser, psmisc, lsb-base, openjdk-8-jre | java8-runtime
Conflicts:
```

spacecowboy commented:
## Upgrade test results

### 14.04 (java8 is not available)

Unable to upgrade due to lack of java 8. `apt-get install -f` will
remove neo4j.

```
root@20ded745fa7a:/workspace# dpkg -i neo4j_3.0.5.SNAPSHOT_all.deb
(Reading database ... 17677 files and directories currently installed.)
Preparing to unpack neo4j_3.0.5.SNAPSHOT_all.deb ...
invoke-rc.d: policy-rc.d denied execution of stop.
Moving obsolete conffile /etc/init.d/neo4j-service out of the way...
Moving obsolete conffile /etc/neo4j/README.txt out of the way...
Moving obsolete conffile /etc/neo4j/neo4j-http-logging.xml out of the way...
Moving obsolete conffile /etc/neo4j/neo4j-server.properties out of the way...
Moving obsolete conffile /etc/neo4j/neo4j.properties out of the way...
Unpacking neo4j (3.0.5.SNAPSHOT) over (2.2.11.SNAPSHOT) ...
dpkg: dependency problems prevent configuration of neo4j:
 neo4j depends on openjdk-8-jre | java8-runtime; however:
  Package openjdk-8-jre is not installed.
  Package java8-runtime is not installed.

dpkg: error processing package neo4j (--install):
 dependency problems - leaving unconfigured
Processing triggers for ureadahead (0.100.0-16) ...
Errors were encountered while processing:
 neo4j
```

### 16.04 (java8 is available but java7 is not, so java8 is already the installed version)

No additional packages are installed, neo4j upgrades correctly.

### Jessie (java7 and java8 are available if backports configured)

neo4j 2.x had java7 installed so that is installed when upgrading to 3.0.

After upgrade, both java7 and java8 are installed but java7 remains
the default and neo4j can't start. Necessary to either uninstall java7
or do:

```
update-java-alternatives --jre -s java-1.8.0-openjdk-amd64
```

This is not a change from previous behavior but I'm not sure anyone
has actually noted it before.
